Add translations to everything

CendioOssman opened the pull request:
Currently, only the native viewer is translated. However, a lot of text the user sees also comes from the common library code.

There are also other components in the project that should be translated for users.

This PR moves the translation infrastructure to the common libraries, and tags everything for translation.

CendioOssman committed:
Change p_() to C_() for translations with context

p_() is commonly used for plural, whilst C_() is what's used by
e.g. glib (and hence Gnome). Let's try to follow what most people
are doing.

CendioOssman committed:
Move translation handling to common/core

This will allow us to translate messages from the common library
as well.

CendioOssman committed:
Include i18n.h where v*printf() functions are used

The translations might use POSIX extensions to the format strings, which
means we may need the gettext versions of the printf() functions where
we don't directly control the format string.

CendioOssman committed:
Add gettext NC_() alias

We don't use this yet, but let's add it so we have the exact same set as
glib, to keep things familiar.

CendioOssman committed:
Add support for plural form gettext

We don't use these yet, but let's make sure the infrastructure is in
place if we want to use it in the future. This should mean we have
complete gettext support.

CendioOssman committed:
Implicitly initialize translations

Make sure we don't have any race between setting up translations and
using them by having them be implicitly loaded on first use.

CendioOssman committed:
Add translations for all exceptions

Match what we are already doing in vncviewer.

Messages that are for internal programming bugs are left untranslated as
they shouldn't really be shown unless something is broken. Avoid
burdening the translators for such theoretical stuff.

CendioOssman committed:
Add translations for parameter descriptions

CendioOssman committed:
Translate placeholders

CendioOssman committed:
Add missing const specifier on auth_error()

We have no intention of modifying this string.

CendioOssman committed:
Handle ISO 8859-1 authentication reason strings

These have historically been ASCII, but let's be prepared for more.

CendioOssman committed:
Translate reasons for rejecting clients

CendioOssman committed:
Translate WinVNC UI

CendioOssman committed:
Ship translations for WinVNC

Now that we translate everything, we need to make sure the translations
are included.

CendioOssman commented:
One issue here is that the translation files will be shared between the viewer and the servers. That's a problem because both the RPMs and debs have things split up in separate packages.

@JoachimFalk and @grulja, do you have any suggestions or preferences here? Any other packages with the same structure that we can mimic?

grulja commented:
Hi, I don't think that would be a problem from the packaging point of view. There are two options for Fedora/RHEL:
1) Currently there is `tigervnc` package with `vncviewer` only. We can make it a separate package, like `tigervnc-viewer` or similar and combine it with currently separate `tigervnc-icons` and `tigervnc-license`. This package would be then required by both `tigervnc-server` and `tigervnc-viewer`.
2) Introduce a new package, something like `tigervnc-common` with the translations, icons and the license and also do the same requires for `tigervnc` (where we have the viewer) and `tigervnc-server`.

I don't know any other package that would have to deal with similar problem, but I can imagine it's done in similar fashion as I wrote above.

JoachimFalk commented:
At the moment, translation are only contained in tigervnc-viewer. But, in Debian, we have tigervnc-common, which tigervnc-tools (vncpasswd, etc.), tigervnc-scraping-server, tigervnc-standalone-server, tigervnc-xorg-extension, and tigervnc-viewer all depend upon.

The common translation would probably go into tigervnc-common if you do not intend to also provide a common library package  for which these translations are and which you also considered to be user consumable, i.e., has public headers and you would consider to be usefull to others. At the moment, the common library code is not made available as a package or even installed as a static lib in the client or server packages.

CendioOssman committed:
Remove barely used LogWriter::status()

It's not clear when this should be used and not info(), so let's just
remove the confusion.

CendioOssman committed:
Merge -license and -icons packages to a -common package

Let's reduce the number of packages we have a bit by merging the shared
stuff.

CendioOssman committed:
Package translations in a shared package

Since every component of TigerVNC is now translated, not just the
viewer.

CendioOssman committed:
Translate vncconfig GUI

CendioOssman committed:
Harmonise some logging strings

To make life easier for the translators.

CendioOssman committed:
Accept long forms of --help/--version for vncconfig

To make things more familiar for users.

CendioOssman committed:
Generate description for alias parameters

This follows a standard pattern, so let's avoid having to duplicate it.

CendioOssman committed:
Copy name and description strings for parameters

They might be generated, so we need to make sure the parameter object
retains its own copy of them.

CendioOssman committed:
Harmonize file I/O error messages

To make things more consistent, and make it easier for the translators.

CendioOssman committed:
Add translations for all default log messages

Match what we are already doing in vncviewer.

Messages that are for internal programming bugs are left untranslated as
they shouldn't really be shown unless something is broken. Avoid
burdening the translators for such theoretical stuff.

CendioOssman committed:
Do Debian installation to temporary directory

This is the proper way of doing things, and will allow us to keep better
track that we get the right set of files in each binary package.

CendioOssman committed:
Fail on unpackaged Ubuntu files

Make sure we don't overlook any files in the future.

CendioOssman committed:
Follow Ubuntu naming for w0vncserver

Set up an alternatives system for this as well, so that we follow the
same pattern as all our other binaries.

CendioOssman committed:
Move translation and metainfo to Ubuntu client package

These were incorrectly included with the server, even though they are
only used by the client. This was a result of not explicitly specifying
what should be in the server package, so all new things ended up there.

CendioOssman committed:
Consistently include config.h

CendioOssman committed:
Actually link libcore.la to Xvnc/libvnc.so

This should have been part of 1c512f7dc. It happened to work because the
other libraries would pull it in as a dependency. But make would be
unable to properly track it for when to rebuild stuff.

CendioOssman committed:
Remove unused OS_LIB variable

It should have been removed in 4001e75.

CendioOssman committed:
Implicitly run setlocale() as needed

We cannot wait for main() to do it, so do it implicitly the first time
it is needed. Try to minimize the effect by only affecting LC_MESSAGES.

CendioOssman committed:
Set up translation helper for Xvnc/libvnc.so

Make sure this is reachable from the C code, and not just the C++ files.

CendioOssman committed:
Always use UTF-8 for translated texts

We assume this encoding for all internal strings, so let's make sure we
never get anything else.

CendioOssman committed:
Support UTF-8 strings in vncconfig GUI
